File: algorithm/go/firstround/142.go
package firstround

type ListNode struct {
	Val  int
	Next *ListNode
}

func hasCycle(head *ListNode) (*ListNode, bool) {
	slow, fast := head, head

	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
		if slow == fast {
			return slow, true
		}
	}

	return nil, false
}

func detectCycle(head *ListNode) *ListNode {
	// 先判断是否有环, 有环找到快慢指针在环中相遇的节点
	p1, isCycle := hasCycle(head)

	// 无环
	if !isCycle {
		return nil
	}

	// 令一个指针指向链表头节点, 让p1 p2以相同速度前进, 当他们再次相遇时就是环的入口节点
	p2 := head
	for p1 != p2 {
		p1 = p1.Next
		p2 = p2.Next
	}

	return p1
}

File: algorithm/cpp/firstround/142.cpp
#include <iostream>
#include <string>
#include <stdio.h>
using namespace std;

struct ListNode
{
    int val;
    ListNode* next;
    ListNode(int x=0): val(x), next(nullptr) {}
};

class Solution
{
public:
    ListNode *detectCycle(ListNode *head);
};

ListNode* Solution::detectCycle(ListNode *head)
{
    if(head==nullptr || head->next==nullptr) return nullptr;

    ListNode* slow = head;
    ListNode* fast = head;

    while (fast!=nullptr && fast->next!=nullptr)
    {
        slow = slow->next;
        fast = fast->next->next;

        if(slow==fast) break;
    }

    if(fast==nullptr || fast->next==nullptr) return nullptr;

    ListNode* p1 = head;
    ListNode* p2 = slow;
    while (p1!=p2)
    {
        p1=p1->next;
        p2=p2->next;
    }

    return p1;
}

int main()
{
    return 0;
}